Mathlib/CategoryTheory/FiberedCategory/Fibered.lean

```diff
@@ -0,0 +1,336 @@
+that for every morphism `f : S ⟶ R` in `𝒮` and every object `b` in `𝒳` with `p(b)=R` there is a
+cartesian arrow `φ : a ⟶ b`  of `f`.
+
+## Implementation
+The standard constructors of `IsCartesian` and `IsFibered` have both been renamed to `.mk'`. We have
+provided alternate lemmas `IsCartesian.mk` and `IsFibered.mk` for constructing instances of these
+structures, and it is recommended to use these instead to minimize the amount of equalities that
+needs to be carried around in the construction.
+
+The reason for this is the following:
+Just like `IsHomLift p f φ`, we have phrased `IsCartesian p f φ` in a way to make its usage as
+flexible  as possible with respect to non-definitional equalities of domains / codomains.
+In particular, given a lift
+```
+  a --φ--> b
+  -        -
+  |        |
+  v        v
+  R --f--> S
+```
+(by which we mean an object of `IsHomLift p f φ`). We say that it is cartesian if for all arrows
+`g : R' ⟶ R`, and all lifts
+```
+  a' --φ'--> b
+  -          -
+  |          |
+  v          v
+  R' --f'--> S
+```
+such that `f' = g ≫ f`, there is a unique induced map `τ : a' ⟶ a` lifting `g` and such that
+`τ ≫ φ = φ'`. This definition gives us some flexibility in that it allows us to take `f'` to be
+non-definitionally equal to `g ≫ f`, and `p(a')` to be non-definitionally equal to `R'`.
+`IsCartesian.mk` only requires us to check this condition for `f' = g ≫ f` and `R=p(a')`.
+
+Similarly, `IsFibered p` is phrased as saying that for every `f : R ⟶ S`, and every `a` such that
+`p(a)=S`, there is a cartesian arrow `φ` lying over `f`. The alternate constructor `IsFibered.mk`
+only requires us to construct this arrow for every `a` and every `f : R ⟶ p(a)`.
+-/
+
+universe v₁ v₂ u₁ u₂
+
+open CategoryTheory Functor Category IsHomLift
+
+variable {𝒮 : Type u₁} {𝒳 : Type u₂} [Category.{v₁} 𝒮] [Category.{v₂} 𝒳]
+
+/-- The proposition that a lift
+```
+  a --φ--> b
+  -        -
+  |        |
+  v        v
+  R --f--> S
+```
+is a cartesian arrow.
+-/
+class IsCartesian (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) extends
+    IsHomLift p f φ : Prop where mk' ::
+  universal_property {R' : 𝒮} {a' : 𝒳} {g : R' ⟶ R} {f' : R' ⟶ S}
+    (_ : f' = g ≫ f) {φ' : a' ⟶ b} (_ : IsHomLift p f' φ') :
+      ∃! χ : a' ⟶ a, IsHomLift p g χ ∧ χ ≫ φ = φ'
+
+protected lemma IsCartesian.mk {p : 𝒳 ⥤ 𝒮} {R S : 𝒮} {a b : 𝒳} {f : R ⟶ S} {φ : b ⟶ a}
+    [IsHomLift p f φ] (h : ∀ (a' : 𝒳) (g : p.obj a' ⟶ R) (φ' : a' ⟶ a), IsHomLift p (g ≫ f) φ' →
+      ∃! χ : a' ⟶ b, IsHomLift p g χ ∧ χ ≫ φ = φ') : IsCartesian p f φ where
+  universal_property := by
+    intro R' a' g f' hf' φ' hφ'
+    have := hφ'.domain_eq.symm
+    subst this
+    subst hf'
+    apply h a' g φ' hφ'
+
+instance {p : 𝒳 ⥤ 𝒮} {R S : 𝒮} {a b : 𝒳} {f : R ⟶ S} {φ : a ⟶ b} [hφ : IsCartesian p f φ] :
+    IsHomLift p f φ := hφ.toIsHomLift
+
+/-- Definition of a Fibered category. -/
+class IsFibered (p : 𝒳 ⥤ 𝒮) : Prop where mk' ::
```

Comment:
This is not the definition of a fibered category, this is a *pre*fibered category. Fibered category are prefibered category where the composition of two cartesian morphisms is cartesian, see SGA 1 VI 6.1.

Reply:
I am using a different notion of `IsCartesian` to SGA, the fact that composition of two cartesian arrows is cartesian follows and is lemma `IsCartesian.comp`. It seems they only require the universal property to be satisfied for `R'=R` (with notation as in the code). Most books on stacks that I have read define them the way we did here, for example Olsson's book. Do you think it's better to define cartesian arrows as in SGA (and then adjust the definition of `IsFibered`)?

Reply:
The notion of prefibered category appears in certain results in the homotopy theory of categories (how certain properties of functors become homotopy equivalences, fibrations, etc, after passing to the nerve), typically Quillen's higher algebraic K-theory. Then, I feel it would be worth defining the notion of prefibered category also. (What you call "cartesian", it is what is named "strongly cartesian" at https://stacks.math.columbia.edu/tag/02XJ ?) For fibered categories, different constructors could be introduced depending on whether we check the expected properties using cartesian or strongly cartesian morphisms.

Reply:
Okay, then I will work to do everything in this generality!

Reply:
I have added this now (although there is still some cleaning up & documentation to write). I think I will split this PR into two: `Cartesian` and `Fibered` once it is ready again.